api/api-cache.php

```diff
@@ -0,0 +1,177 @@
+<?php
+require_once('../template/top.php');
+
+// RegEx to match certain endpoints we want to ignore, e.g., orders
+const IGNORED_ENDPOINTS = '/(?:orders)/i';
+/**
+ * Gets the first entry from the API cache that has a matching endpoint.
+ * @param string $endpoint The endpoint for the API request
+ * @return array|null The first entry that matches or null if no entry exists
+ */
+function get_cached_api_response(string $endpoint): ?array
+{
+    global $db;
+    $q = $db->query('SELECT * FROM api_cache WHERE endpoint = "' . $db->real_escape_string($endpoint) . '"');
+    if ($q) {
+        $r = $q->fetch_array(MYSQLI_ASSOC);
+        return $r;
+    }
+    return null;
+}
+
+class CacheResult{
+    /**
+     * @var string $content The response content
+     */
+    public $content;
+    /**
+     * @var bool $curl_executed true if curl_exec was called, false otherwise
+     */
+    public $curl_executed;
+    /**
+     * @var int|null $httpcode HTTP response code from cURL. Null if curl_exec wasn't called
+     * @var int|null $curl_errno cURL error number. Null if curl_exec wasn't called
+     */
+    public $httpcode, $curl_errno;
+
+    /**
+     * @param string $content
+     * @param bool $curl_executed Defaults to false
+     * @param int|null $httpcode (optional)
+     * @param int|null $curl_errno (optional)
+     */
+    public function __construct(string $content, bool $curl_executed = false, $httpcode = null, $curl_errno = null)
+    {
+        $this->content = $content;
+        $this->curl_executed = $curl_executed;
+        $this->httpcode = $httpcode;
+        $this->curl_errno = $curl_errno;
+    }
+}
+
+/**
+ * Gets the first entry from the API cache that has a matching endpoint. Will not return if the entry is expired.
+ * @param string $endpoint The endpoint for the API request
+ * @param resource|CurlHandle $ch The Curl Handle to send the request
+ * @param int|null $configId (optional) The config ID for the cached response. Optional if result won't be cached
+ * @return CacheResult An object containing the response content, and some cURL info if curl_exec was called
+ */
+function get_valid_cache_entry(string $endpoint, $ch, $configId = null)
+{
+    global $db;
+    if (preg_match(IGNORED_ENDPOINTS, $endpoint)) {
+        return new CacheResult(curl_exec($ch),true,curl_getinfo($ch,CURLINFO_HTTP_CODE),curl_errno($ch));
+    }
+    $q = $db->query("SELECT
+                                * 
+                            FROM 
+                                api_cache
+                            JOIN 
```

Comment:
This might need to be a `LEFT JOIN`, a standard `JOIN` I think will return nothing at all if there's no entry in the `api_cache` table but there is an entry in the `outgoing_request_cache_config` table (i.e. JOIN expects both to exist to return anything, but the first time a cache is ever called there won't be an `api_cache` entry just yet)

Reply:
Changed in c289fad